main.py

```diff
@@ -0,0 +1,273 @@
+    params = {
+        'input_token':  access_token,
+        'access_token': app_token
+    }
+    r = urlfetch.fetch('https://graph.facebook.com/debug_token?' +
+                       urllib.urlencode(params))
+    result = json.loads(r.content)
+
+    # If the response includes `is_valid` being false, fail
+    if result['data']['is_valid'] is False:
+        return make_response('Authentication failed.', 401)
+
+    # Make an API request to Facebook using OAuth
+    r = urlfetch.fetch('https://graph.facebook.com/me?fields=name,email',
+                       headers={'Authorization': 'OAuth '+access_token})
+    idinfo = json.loads(r.content)
+
+    # Save the Facebook profile
+    store = CredentialStore(id=idinfo['id'], profile=idinfo)
+    store.put()
+
+    # Obtain the Facebook user's image
+    profile = idinfo
+    profile['imageUrl'] = 'https://graph.facebook.com/' + profile['id'] +\
+        '/picture?width=96&height=96'
+
+    # Not making a session for demo purpose/simplicity
+    return make_response(json.dumps(profile), 200)
+
+
+@app.route('/register', methods=['POST'])
+def register():
+    # The POST should include `email`
+    email = request.form.get('email', None)[:32]
+
+    # The POST should include `password`
+    _password = request.form.get('password', None)[:32]
+
+    # Validate the parameters POST'ed (intentionally not too strict)
+    if not email or not _password:
+        return make_response('Bad Request', 400)
+
+    # Hash password
+    password = CredentialStore.hash(_password)
+    # Perform relevant sanitization/validation on your own code.
+    # This demo omits them on purpose for simplicity.
+    profile = {
+        'id':       email,
+        'email':    email,
+        'name':     request.form.get('name', ''),
+        'password': password,
+        'imageUrl': ''
+    }
+
+    # Overwrite existing user
+    store = CredentialStore(id=profile['id'], profile=profile)
+    store.put()
+
+    # Get rid of password from profile
+    profile.pop('password')
+
+    # Not making a session for demo purpose/simplicity
+    return make_response(json.dumps(profile), 200)
+
+
+@app.route('/unregister', methods=['POST'])
+def unregister():
+    if 'id' not in request.form:
+        make_response('User id not specified', 400)
+
+    id = request.form.get('id', '')
+    store = CredentialStore.get_by_id(str(id))
+
+    if store is None:
+        make_response('User not registered', 400)
+
+    profile = store.profile
+
+    if profile is None:
+        return make_response('Failed', 400)
+
+    # Remove the user account
+    CredentialStore.remove(str(id))
+    # Not terminating a session for demo purpose/simplicity
+    return make_response('Success', 200)
+
+
+@app.route('/signout', methods=['POST'])
+def signout():
+    # Not terminating a session for demo purpose/simplicity
+    return make_response(json.dumps({}), 200)
+
+
+@app.route('/', defaults={'path': ''})
+@app.route('/<path>')
```

Comment:
After some discussion, I think we would like to avoid adding this server-side dependency to the demo. There's nothing wrong with it, but we advise folks to use the PRPL pattern and app engine doesn't have good support for H2 as of last time we tried. Would it be possible to just support Google Sign In by using the [Google Sign In element](https://elements.polymer-project.org/elements/google-signin?view=demo:demo/index.html&active=google-signin)?